File: server/internal/routes/sync.go
package routes

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
	"gorm.io/gorm"
	"notes/server/internal/middleware"
	"notes/server/internal/schemas"
	"notes/server/internal/services"
)

type SyncHandler struct {
	service services.SyncService
}

func RegisterSyncRoutes(r chi.Router, syncService services.SyncService, jwtSecret string, conn *gorm.DB) {
	handler := SyncHandler{service: syncService}

	r.Route("/api/sync", func(r chi.Router) {
		r.Use(middleware.AuthMiddleware(jwtSecret, conn))
		r.Post("/", handler.sync)
	})
}

func (h SyncHandler) sync(w http.ResponseWriter, r *http.Request) {
	userID := middleware.UserIDFromContext(r.Context())
	if userID == "" {
		middleware.WriteAuthError(w)
		return
	}

	var req schemas.SyncRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid json payload", http.StatusBadRequest)
		return
	}

	resp, err := h.service.Sync(userID, req)
	if err != nil {
		http.Error(w, "sync failed: "+err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(resp)
}

File: server/routes/sync.ts
import type { FastifyInstance } from 'fastify';
import { authMiddleware } from '@middleware/auth';
import { sync } from '@services/sync';
import type { SyncRequest } from '@types/index';

export async function syncRoutes(fastify: FastifyInstance) {
  fastify.post<{ Body: SyncRequest }>(
    '/sync',
    { preHandler: authMiddleware },
    async (request, reply) => {
      try {
        const userId = (request as any).userId as string;
        const body = request.body;

        if (!Array.isArray(body?.ops)) {
          return reply.status(400).send({ error: 'ops must be an array' });
        }

        const result = await sync(userId, body);
        return reply.send(result);
      } catch (err: any) {
        fastify.log.error(err);
        return reply.status(500).send({ error: 'Sync failed', message: err.message });
      }
    }
  );
}
